File: +ndi/+setup/+NDIMaker/epochProbeMapMaker.m
% Folder: +ndi/+setup/+NDIMaker/
classdef epochProbeMapMaker < handle
    %EPOCHPROBEMAPMAKER Creates or updates epoch probe map files for NDI sessions.
    %   The EPOCHPROBEMAPMAKER class automates the generation of '.epochprobemap.txt'
    %   files. These files map experimental epochs to data acquisition probes and devices.
    %   The class takes a base path (typically an NDI session directory), a table
    %   defining epoch-specific variables, and a table defining probe characteristics.
    %   It processes these inputs to create or update the epoch probe map files,
    %   associating subject information and device strings with each epoch and probe.
    %   The class includes options for validating input data and handling existing
    %   map files.

    properties (Access = public)
        path (1,:) char         % Base directory path where session folders are located or will be created.
        variableTable table     % Input table containing session definition information. Must contain 'SubjectString' and 'SessionPath' variables.
        probeTable table        % Input table containing session definition information. Must contain 'SubjectString' and 'SessionPath' variables.
    end

    methods
        function obj = epochProbeMapMaker(path,variableTable,probeTable,options)
            %EPOCHPROBEMAPMAKER Constructor for the epochProbeMapMaker class.
            %   OBJ = EPOCHPROBEMAPMAKER(PATH, VARIABLETABLE, PROBETABLE) creates an
            %   epochProbeMapMaker object and generates '.epochprobemap.txt' files.
            %   It uses the 'RowNames' from VARIABLETABLE as base filenames for the
            %   map files, which are saved in the directory specified by PATH.
            %   Each map file links probes defined in PROBETABLE to the corresponding
            %   epoch, using subject information from VARIABLETABLE.
            %
            %   OBJ = EPOCHPROBEMAPMAKER(..., 'Name', Value) allows specifying
            %   additional options to control the map creation process.
            %
            %   Input Arguments:
            %       path            - The absolute path to an NDI session directory
            %                           The generated '.epochprobemap.txt' files will 
            %                           be saved directly within this directory.
            %                           Must be an existing folder.
            %       variableTable   - A MATLAB table defining epoch-specific variables.
            %                           - Its `RowNames` must be unique epoch identifiers
            %                             which will serve as base filenames for the 
            %                             output '.epochprobemap.txt' files.
            %                           - Must contain a 'SubjectString' column (char or string).
            %                           - Other columns can be included and used for
            %                             validation via the 'NonNaNVariableNames' option.
            %       probeTable      - A MATLAB table defining probe characteristics. It 
            %                           must contain the following columns:
            %                           - 'name' (cellstr or string array): Name of the probe.
            %                           - 'reference' (numeric array): Reference number of the probe.
            %                           - 'type' (cellstr or string array): Type of the probe.
            %                           - 'deviceString' (cellstr or string array): Device string
            %                             associated with the probe.
            %
            %   Optional Name-Value Arguments:
            %       Overwrite        - If true, existing epoch probe maps found at the specified 
            %                           paths will be overwritten. Default: false.
            %       NonNaNVariableNames - Variable names in 'variableTable'.Values in 
            %                           these columns must not be NaN for a valid 
            %                           session to be created. Default: {}.
            %       ProbePostfix     - A postfix to be appended to probe names. Default: {} (no postfix).
            %                           - If a char/string:
            %                               - If a variable name in variableTable, the
            %                                   value of that variable in the corresponding
            %                                   row is used.
            %                               - If not a variable name, this char/string is used.
            %                           - If a cell array: 
            %                               - If its length matches the total number of rows
            %                                   in the input `variableTable`, the value 
            %                                   in the corresponding row is used.
            %                               - If its length matches the total number of rows
            %                                   in the input `variableTable` and the second dimension
            %                                   matches the number of probes (i.e. height of probeTable),
            %                                   the value in the corresponding row is used for each probe.
            %                               - If values are variable names with length
            %                                   equal to the number of probes (i.e. height of
            %                                   probeTable, the value of the variable(s) 
            %                                   in the corresponding row are used. An empty
            %                                   char/str vector may be used to omit a postfix for
            %                                   certain variables (e.g. {'PostFix','','PostFix'})
            %                           
            %   Output Arguments:
            %       obj (epochProbeMapMaker)  - The constructed epochProbeMapMaker object.
            %
            %   See also: NDI.EPOCH.EPOCHPROBEMAP_DAQSYSTEM

            arguments
                path (1,:) char
                variableTable table
                probeTable table
                options.Overwrite (1,1) logical = false;
                options.NonNaNVariableNames {mustBeA(options.NonNaNVariableNames,{'char','str','cell'})} = {};
                options.ProbePostfix {mustBeA(options.ProbePostfix,{'char','str','cell'})} = {};
            end

             % Assign properties from inputs
            obj.path = path;
            obj.variableTable = variableTable;
            obj.probeTable = probeTable;

            % Create progress bar
            progressBar = ndi.gui.component.ProgressBarWindow('Import Dataset','Overwrite',false);
            progressBar = progressBar.addBar('Label','Creating Epoch Probe Map(s)','Tag','epochprobemap');

            % Validate variableTable: check for required 'SubjectString' column
            if ~ismember('SubjectString', variableTable.Properties.VariableNames)
                error('epochProbeMapMaker:MissingSubjectString', ...
                    "The 'variableTable' must contain a 'SubjectString' column.");
            end

            % Validate probeTable: check for required columns
            requiredProbeCols = {'name', 'reference', 'type', 'deviceString'};
            missingProbeCols = setdiff(requiredProbeCols, probeTable.Properties.VariableNames);
            if ~isempty(missingProbeCols)
                error('epochProbeMapMaker:MissingProbeTableColumns', ...
                    "The 'probeTable' is missing the following required columns: %s.", strjoin(missingProbeCols, ', '));
            end

            % Get valid epoch rows
            validInd = find(ndi.util.identifyValidRows(variableTable,options.NonNaNVariableNames));
            if isempty(validInd)
                warning('epochProbeMapMaker:NoValidEpochs', ...
                    'No valid epochs found in variableTable after NaN checks. No epochprobemaps will be created.');
                return;
            end
            
            % Get epoch data filenames
            epochstreams = fullfile(path,variableTable.Properties.RowNames(validInd));
            
            for e = 1:numel(epochstreams)

                % Construct the full filename for the epoch probe map file
                [pathname,filename] = fileparts(epochstreams{e});
                probeFilename = fullfile(pathname,strcat(filename,'.epochprobemap.txt'));
                
                % Skip if not overwriting and file exists
                if ~options.Overwrite && exist(probeFilename, 'file')
                    continue
                end
                
                % Initialize an array of NDI epochprobemap_daqsystem objects for the current epoch
                probemap = ndi.epoch.epochprobemap_daqsystem.empty(0,height(probeTable));

                for p = 1:height(probeTable)

                    probeName = probeTable.name{p}; % Original probe name
                    
                    if ischar(options.ProbePostfix)
                        % Check if Postfix matches a variable name in variableTable
                        if any(strcmpi(variableTable.Properties.VariableNames,options.ProbePostfix))
                            probeName = strcat(probeName,variableTable.(options.ProbePostfix){validInd(e)});
                        
                        % If not, just append Postfix as is
                        else
                            probeName = strcat(probeName,options.ProbePostfix);
                        end
                    elseif iscell(options.ProbePostfix)
                        % Check if Postfix is the same length as the variableTable
                        if numel(options.ProbePostfix) == height(variableTable)
                            probeName = strcat(probeName,options.ProbePostfix{validInd(e)});

                        % Check if Postfix size matches the height(variableTable) x height(probeTable)
                        elseif size(options.ProbePostfix) == [height(variableTable),height(probeTable)]
                            probeName = strcat(probeName,options.ProbePostfix{validInd(e),p});

                        % Check if Postfix size matches the height(probeTable) x height(variableTable)
                        elseif size(options.ProbePostfix) == [height(probeTable),height(variableTable)]
                            probeName = strcat(probeName,options.ProbePostfix{p,validInd(e)});
                        
                        % Check if Postfix contains variable names in variableTable
                        elseif numel(options.ProbePostfix) == height(probeTable)
                            if isempty(options.ProbePostfix{p})
                                probeName = probeName;
                            elseif any(strcmpi(variableTable.Properties.VariableNames,options.ProbePostfix{p}))
                                probeName = strcat(probeName,variableTable.(options.ProbePostfix{p}){validInd(e)});
                            else
                                 warning('EPOCHPROBEMAPMAKER:InvalidPostfix',...
                                    'ProbePostfix %s is not valid.',options.ProbePostfix{p})
                            end
                        else
                            warning('EPOCHPROBEMAPMAKER:InvalidPostfix',...
                                'ProbePostfix is not valid.')
                        end
                    else
                        warning('EPOCHPROBEMAPMAKER:InvalidPostfix',...
                                'ProbePostfix is not valid.')
                    end

                    % Create an epochprobemap_daqsystem object for the current probe
                    probemap(p) = ndi.epoch.epochprobemap_daqsystem(...
                        probeName,...
                        probeTable.reference{p},...
                        probeTable.type{p},...
                        probeTable.deviceString{p},...
                        variableTable.SubjectString{validInd(e)});
                end

                % Save the NDI epochprobemap objects to the file.
                probemap.savetofile(probeFilename);

                % Update progress bar
                progressBar = progressBar.updateBar('epochprobemap',e/numel(epochstreams));
                if strcmpi(progressBar.getStatus('epochprobemap'),'Pause')
                    uiwait(progressBar.ProgressFigure);
                end
            end
        end
    end
end

File: ndi-python/ndi/setup/makers/epoch_probe_map_maker.py
"""
NDI Epoch Probe Map Maker - Create epoch probe map files for NDI sessions.

This module provides EpochProbeMapMaker class for automating the generation
of '.epochprobemap.txt' files that map experimental epochs to data acquisition
probes and devices.
"""

from typing import List, Dict, Any, Optional, Union
import os
import logging
import pandas as pd
import numpy as np


class EpochProbeMapMaker:
    """
    Creates or updates epoch probe map files for NDI sessions.

    The EpochProbeMapMaker class automates the generation of '.epochprobemap.txt'
    files. These files map experimental epochs to data acquisition probes and devices.
    The class takes a base path (typically an NDI session directory), a table
    defining epoch-specific variables, and a table defining probe characteristics.

    Attributes:
        path (str): Base directory path where session folders are located.
        variable_table (pd.DataFrame): Input table containing epoch definition information.
                                       Must contain 'SubjectString' column and row names.
        probe_table (pd.DataFrame): Input table containing probe characteristics.
                                   Must contain 'name', 'reference', 'type', 'deviceString' columns.

    Example:
        >>> import pandas as pd
        >>> from ndi.setup.makers import EpochProbeMapMaker
        >>>
        >>> # Create variable table with epoch information
        >>> var_data = {
        ...     'SubjectString': ['subject_001', 'subject_001'],
        ... }
        >>> variable_table = pd.DataFrame(var_data, index=['epoch1', 'epoch2'])
        >>>
        >>> # Create probe table
        >>> probe_data = {
        ...     'name': ['probe1', 'probe2'],
        ...     'reference': [1, 2],
        ...     'type': ['electrode', 'electrode'],
        ...     'deviceString': ['dev1', 'dev2']
        ... }
        >>> probe_table = pd.DataFrame(probe_data)
        >>>
        >>> # Create epoch probe maps
        >>> maker = EpochProbeMapMaker('/path/to/session', variable_table, probe_table)
    """

    def __init__(self,
                 path: str,
                 variable_table: pd.DataFrame,
                 probe_table: pd.DataFrame,
                 overwrite: bool = False,
                 non_nan_variable_names: Optional[List[str]] = None,
                 probe_postfix: Union[str, List[str], None] = None,
                 show_progress: bool = True):
        """
        Initialize EpochProbeMapMaker and create epoch probe map files.

        Args:
            path: The absolute path to an NDI session directory where
                  '.epochprobemap.txt' files will be saved.
            variable_table: A DataFrame defining epoch-specific variables.
                           Index values serve as base filenames for output files.
                           Must contain a 'SubjectString' column.
            probe_table: A DataFrame defining probe characteristics. Must contain:
                        - 'name': Name of the probe
                        - 'reference': Reference number of the probe
                        - 'type': Type of the probe
                        - 'deviceString': Device string associated with the probe
            overwrite: If True, existing epoch probe maps will be overwritten.
                      Default: False.
            non_nan_variable_names: Column names in variable_table whose values
                                   must not be NaN for valid epoch. Default: None.
            probe_postfix: Optional postfix to append to probe names. Can be:
                          - String: Applied to all probes
                          - List[str]: One per probe or per epoch
                          Default: None.
            show_progress: Whether to display progress. Default: True.

        Raises:
            ValueError: If required columns are missing or path doesn't exist.
        """
        from ...util.table import identify_valid_rows
        from ...gui import ProgressTracker, ConsoleProgressMonitor

        # Validate inputs
        if not os.path.isdir(path):
            raise ValueError(f"Path does not exist: {path}")

        if 'SubjectString' not in variable_table.columns:
            raise ValueError("variable_table must contain 'SubjectString' column")

        required_probe_cols = ['name', 'reference', 'type', 'deviceString']
        missing_cols = [col for col in required_probe_cols if col not in probe_table.columns]
        if missing_cols:
            raise ValueError(f"probe_table missing required columns: {', '.join(missing_cols)}")

        # Assign properties
        self.path = path
        self.variable_table = variable_table
        self.probe_table = probe_table

        # Get valid epoch rows
        if non_nan_variable_names is None:
            non_nan_variable_names = []

        valid_ind = identify_valid_rows(variable_table, non_nan_variable_names)

        if not any(valid_ind):
            logging.warning('No valid epochs found in variable_table. No epoch probe maps will be created.')
            return

        # Get valid indices
        valid_indices = np.where(valid_ind)[0]

        # Get epoch identifiers from index (row names)
        if variable_table.index.name is None and all(isinstance(idx, int) for idx in variable_table.index):
            # Default integer index - use row number
            epoch_ids = [f"epoch_{i}" for i in valid_indices]
        else:
            # Use actual index values
            epoch_ids = [variable_table.index[i] for i in valid_indices]

        logging.info(f"Creating epoch probe map files for {len(epoch_ids)} epoch(s)...")

        # Setup progress monitoring
        progress_tracker = None
        progress_monitor = None
        if show_progress and len(epoch_ids) > 0:
            progress_tracker = ProgressTracker()
            progress_monitor = ConsoleProgressMonitor("Creating Epoch Probe Map(s)", tracker=progress_tracker)
            progress_tracker.start()

        # Process each valid epoch
        for idx, epoch_idx in enumerate(valid_indices):
            epoch_id = epoch_ids[idx]

            # Construct the full filename for the epoch probe map file
            probe_filename = os.path.join(path, f"{epoch_id}.epochprobemap.txt")

            # Skip if not overwriting and file exists
            if not overwrite and os.path.exists(probe_filename):
                logging.debug(f"Epoch {idx+1}/{len(epoch_ids)}: Skipping existing '{epoch_id}'")
                continue

            # Create probe map entries for this epoch
            probe_map_entries = []

            for p in range(len(probe_table)):
                probe_name = probe_table.iloc[p]['name']

                # Apply probe postfix if provided
                if probe_postfix is not None:
                    if isinstance(probe_postfix, str):
                        # Check if it's a variable name in variable_table
                        if probe_postfix in variable_table.columns:
                            postfix_value = variable_table.iloc[epoch_idx][probe_postfix]
                            if postfix_value is not None and postfix_value != '':
                                probe_name = f"{probe_name}{postfix_value}"
                        else:
                            # Use as literal string
                            probe_name = f"{probe_name}{probe_postfix}"

                    elif isinstance(probe_postfix, list):
                        # Multiple postfix options
                        if len(probe_postfix) == len(variable_table):
                            # One per epoch
                            postfix_value = probe_postfix[epoch_idx]
                            if postfix_value:
                                probe_name = f"{probe_name}{postfix_value}"
                        elif len(probe_postfix) == len(probe_table):
                            # One per probe
                            postfix_value = probe_postfix[p]
                            if postfix_value:
                                # Check if it's a variable name
                                if postfix_value in variable_table.columns:
                                    var_value = variable_table.iloc[epoch_idx][postfix_value]
                                    if var_value is not None and var_value != '':
                                        probe_name = f"{probe_name}{var_value}"
                                else:
                                    probe_name = f"{probe_name}{postfix_value}"
                        else:
                            logging.warning(f"probe_postfix length ({len(probe_postfix)}) doesn't match "
                                          f"epochs ({len(variable_table)}) or probes ({len(probe_table)})")

                # Get probe information
                probe_ref = probe_table.iloc[p]['reference']
                probe_type = probe_table.iloc[p]['type']
                device_string = probe_table.iloc[p]['deviceString']
                subject_string = variable_table.iloc[epoch_idx]['SubjectString']

                # Create probe map entry
                # Format: name reference type deviceString subjectString
                entry = f"{probe_name}\t{probe_ref}\t{probe_type}\t{device_string}\t{subject_string}"
                probe_map_entries.append(entry)

            # Write the epoch probe map file
            try:
                with open(probe_filename, 'w') as f:
                    f.write('\n'.join(probe_map_entries))
                    f.write('\n')  # End with newline

                logging.info(f"Epoch {idx+1}/{len(epoch_ids)}: Created '{epoch_id}.epochprobemap.txt' "
                           f"with {len(probe_map_entries)} probe(s)")

            except Exception as e:
                logging.error(f"Failed to write epoch probe map file '{probe_filename}': {e}")

            # Update progress
            if progress_tracker:
                progress_tracker.update_progress(
                    (idx + 1) / len(epoch_ids),
                    f"Epoch {idx+1}/{len(epoch_ids)}: {epoch_id}"
                )

        # Mark progress complete
        if progress_tracker:
            progress_tracker.mark_complete()

        logging.info(f"Completed epoch probe map creation for {len(epoch_ids)} epoch(s)")

    def __repr__(self) -> str:
        """String representation of EpochProbeMapMaker."""
        return (f"EpochProbeMapMaker(path='{self.path}', "
                f"epochs={len(self.variable_table)}, "
                f"probes={len(self.probe_table)})")
